src/main/java/com/google/devtools/build/skydoc/rendering/MarkdownUtil.java

```diff
@@ -46,55 +47,131 @@ public MarkdownUtil(String extensionBzlFile) {
-   * This performs the following operations:
+   * Formats the input string so that it is displayable in a Markdown table cell. This performs the
+   * following operations:
    *
    * <ul>
    *   <li>Trims the string of leading/trailing whitespace.
-   *   <li>Transforms the string using {@link #htmlEscape}.
-   *   <li>Transforms multline code (```) tags into preformatted code HTML tags.
-   *   <li>Transforms single-tick code (`) tags into code HTML tags.
-   *   <li>Transforms 'new paraphgraph' patterns (two or more sequential newline characters) into
-   *       line break HTML tags.
-   *   <li>Turns lingering new line tags into spaces (as they generally indicate intended line wrap.
+   *   <li>Escapes pipe characters ({@code |}) as {@code \|}.
+   *   <li>Transforms Markdown code blocks ({@code ```}) into HTML preformatted code blocks, and
+   *       transforms newlines within those code blocks into character entities
+   *   <li>Transforms remaining 'new paragraph' patterns (two or more sequential newline characters)
+   *       into line break HTML tags.
+   *   <li>Turns remaining newlines into spaces (as they generally indicate intended line wrap).
    * </ul>
+   *
+   * TODO(https://github.com/bazelbuild/stardoc/issues/118): also format Markdown lists as HTML.
    */
-  public String markdownCellFormat(String docString) {
-    String resultString = htmlEscape(docString.trim());
+  public static String markdownCellFormat(String docString) {
+    return new MarkdownCellFormatter(docString).format();
+  }
 
-    resultString = replaceWithTag(resultString, "```", "<pre><code>", "</code></pre>");
-    resultString = replaceWithTag(resultString, "`", "<code>", "</code>");
+  // See https://github.github.com/gfm
+  private static final class MarkdownCellFormatter {
+    private final ImmutableList<String> lines;
+    int currentLine;
+    StringBuilder result;
 
-    return resultString.replaceAll("\n(\\s*\n)+", "<br><br>").replace('\n', ' ');
-  }
+    private static final Pattern CODE_BLOCK_OPENING_FENCE =
+        Pattern.compile(" {0,3}(?<fence>```+|~~~+) *(?<lang>\\w*)[^`~]*");
 
-  private static String replaceWithTag(
-      String wholeString, String stringToReplace, String openTag, String closeTag) {
-    String remainingString = wholeString;
-    StringBuilder resultString = new StringBuilder();
+    MarkdownCellFormatter(String docString) {
+      lines = docString.trim().replace("|", "\\|").lines().collect(toImmutableList());
+      currentLine = 0;
+      result = new StringBuilder();
+    }
 
-    boolean openTagNext = true;
-    int index = remainingString.indexOf(stringToReplace);
-    while (index > -1) {
-      resultString.append(remainingString, 0, index);
-      resultString.append(openTagNext ? openTag : closeTag);
-      openTagNext = !openTagNext;
-      remainingString = remainingString.substring(index + stringToReplace.length());
-      index = remainingString.indexOf(stringToReplace);
+    String format() {
```

Comment:
Nit: Javadoc that this consumes the whole input and yields the result.

Reply:
Fixed, good point.